In [1]:
# Imports

import pandas as pd
import numpy as np
import os
import pprint

%matplotlib inline
import matplotlib as plt
import seaborn as sns

In [2]:
# function to see how many nulls are in the columns

def null_counts(df):
    """
    function to count the number of null values in each df column
    """
    missing = df.isna().sum()
    return missing

In [3]:
# function to determine percentage of NaN values in each column

def null_percentages(df):
    """
    returns the percentage of NaN values in each column to help determine
    whether or not the column in question should be kept or dropped
    """
    
    perct = round((df.isna().sum()) / len(df) * 100), 2
    return perct

In [4]:
# function to combine both 'null_counts' and 'null_percentages'

def null_feedback(df):
    """
    function that tells the user the info on the DataFrame's NaNs
    """
    x = null_counts(df)
    y = null_percentages(df)
    print(f"Total number of NaNs by column: \n{x}.")
    print(f"Percentage of columns that are NaN: \n{y}.")
    return 

## Player_data DataFrame Work

In [5]:
# bring in the 'player_data' csv

player_data_df = pd.read_csv("/Users/DataScience/Rimshotz/alley-oop-nba-stats/player_data.csv")

player_data_shape = player_data_df.shape

a = player_data_df.isnull().sum().sum()

print("Player Data Information")
print(f"Consisting of {player_data_shape[0]} rows and {player_data_shape[1]} columns")
print(f"It has loads of data, but also has {a} missing values.") 
player_data_df

Player Data Information
Consisting of 4550 rows and 8 columns
It has loads of data, but also has 341 missing values.


name  year_start  year_end position height  weight  \
0          Alaa Abdelnaby        1991      1995      F-C   6-10   240.0   
1         Zaid Abdul-Aziz        1969      1978      C-F    6-9   235.0   
2     Kareem Abdul-Jabbar        1970      1989        C    7-2   225.0   
3      Mahmoud Abdul-Rauf        1991      2001        G    6-1   162.0   
4       Tariq Abdul-Wahad        1998      2003        F    6-6   223.0   
...                   ...         ...       ...      ...    ...     ...   
4545           Ante Zizic        2018      2018      F-C   6-11   250.0   
4546             Jim Zoet        1983      1983        C    7-1   240.0   
4547            Bill Zopf        1971      1971        G    6-1   170.0   
4548          Ivica Zubac        2017      2018        C    7-1   265.0   
4549           Matt Zunic        1949      1949      G-F    6-3   195.0   

             birth_date                                college  
0         June 24, 1968                        Duke University  
1         April 7, 1946                  Iowa State University  
2        April 16, 1947  University of California, Los Angeles  
3         March 9, 1969             Louisiana State University  
4      November 3, 1974              San Jose State University  
...                 ...                                    ...  
4545    January 4, 1997                                    NaN  
4546  December 20, 1953                  Kent State University  
4547       June 7, 1948                    Duquesne University  
4548     March 18, 1997                                    NaN  
4549  December 19, 1919           George Washington University  

[4550 rows x 8 columns]

**Let's find out where those missing values are located**

In [6]:
player_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4550 entries, 0 to 4549
Data columns (total 8 columns):
name          4550 non-null object
year_start    4550 non-null int64
year_end      4550 non-null int64
position      4549 non-null object
height        4549 non-null object
weight        4544 non-null float64
birth_date    4519 non-null object
college       4248 non-null object
dtypes: float64(1), int64(2), object(5)
memory usage: 284.5+ KB


**Initial observations from data import:**

- Strings where they're expected ("name" and "college" columns), but also where NOT expected ("height" and "birthdate")

- "height" column is written so that someone who is 6-feet, 9-inches tall is documented as '6-9.'  We will have to not only change the data in that column to int/float, but also convert all those heights to inches.  Choosing inches because the representation as it is on import is NOT in metric (m/cm)

- There are some NaNs in the dataset that must be dealt with.  Recall: 'impute' is to assign a value; 'encode' is to give it a 0 or 1 

In [7]:
player_data_df.shape

(4550, 8)

In [8]:
null_feedback(player_data_df)

Total number of NaNs by column: 
name            0
year_start      0
year_end        0
position        1
height          1
weight          6
birth_date     31
college       302
dtype: int64.
Percentage of columns that are NaN: 
(name          0.0
year_start    0.0
year_end      0.0
position      0.0
height        0.0
weight        0.0
birth_date    1.0
college       7.0
dtype: float64, 2).


In [9]:
a = null_counts(player_data_df)
b = null_percentages(player_data_df)

print(f"The Player Data DataFrame has a total of {a} nulls, for a total null pct of {b}.")

The Player Data DataFrame has a total of name            0
year_start      0
year_end        0
position        1
height          1
weight          6
birth_date     31
college       302
dtype: int64 nulls, for a total null pct of (name          0.0
year_start    0.0
year_end      0.0
position      0.0
height        0.0
weight        0.0
birth_date    1.0
college       7.0
dtype: float64, 2).


In [10]:
# from datetime import datetime

# def birth_date(df):
#     """
#     function to loop through all the strings in 'birth_date' and convert
#     them into more data-analysis-friendly datetime formats
#     """
#     for birth_date in player_data_df.birth_date:
#         datetime_object = datetime.strptime('Jan 1 2005', '%b %d %Y')
#     df = birth_date_df
#     return birth_date_df


# player_data_df["birth_date"] = pd.date_range(pd.Timestamp(player_data_df.birth_date),
#                                             periods=1)

# player_data.strftime("%B %d, %Y")

In [11]:
# replace 'height' column with inches

# player_data_df['height'] = player_data_df.height.str.split("-").apply(lambda x: int(x[0]) * 12 + int(x[1]))

# def true_height(strings):
#     for string in strings:
#         new_str = string.split('-')
#         inches1 = new_str[0]
#         inches2 = new_str[1]

#         inches1 = int(inches1)*12
#         inches2 = int(inches2)

#     return inches1 + inches2

# player_data_df['height'] = player_data_df.height.apply(true_height)

# remove the dash from heights
# player_data_df["height"] = player_data_df["height"].str.replace("-", " ")

# player_data_df["height"]=player_data_df["height"].replace(["-"], "")
player_data_df

name  year_start  year_end position height  weight  \
0          Alaa Abdelnaby        1991      1995      F-C   6-10   240.0   
1         Zaid Abdul-Aziz        1969      1978      C-F    6-9   235.0   
2     Kareem Abdul-Jabbar        1970      1989        C    7-2   225.0   
3      Mahmoud Abdul-Rauf        1991      2001        G    6-1   162.0   
4       Tariq Abdul-Wahad        1998      2003        F    6-6   223.0   
...                   ...         ...       ...      ...    ...     ...   
4545           Ante Zizic        2018      2018      F-C   6-11   250.0   
4546             Jim Zoet        1983      1983        C    7-1   240.0   
4547            Bill Zopf        1971      1971        G    6-1   170.0   
4548          Ivica Zubac        2017      2018        C    7-1   265.0   
4549           Matt Zunic        1949      1949      G-F    6-3   195.0   

             birth_date                                college  
0         June 24, 1968                        Duke University  
1         April 7, 1946                  Iowa State University  
2        April 16, 1947  University of California, Los Angeles  
3         March 9, 1969             Louisiana State University  
4      November 3, 1974              San Jose State University  
...                 ...                                    ...  
4545    January 4, 1997                                    NaN  
4546  December 20, 1953                  Kent State University  
4547       June 7, 1948                    Duquesne University  
4548     March 18, 1997                                    NaN  
4549  December 19, 1919           George Washington University  

[4550 rows x 8 columns]

In [12]:
# find the NaN value in the df
player_data_df[player_data_df["height"].isnull()].index.tolist()

[2142]

In [13]:
# show me row 2142

print(player_data_df.loc[[2142]])

             name  year_start  year_end position height  weight    birth_date  \
2142  George Karl        1974      1978      NaN    NaN     NaN  May 12, 1952   

                           college  
2142  University of North Carolina  


**Okay, so we know people lie about their weight, but having a 'NaN' value in the weight column is *killing* my statistical analysis, so substituting George Karl's NaN weight with what I find his playing weight to be while with the San Antonio Spurs from 1974-1978**

- [Team Roster Source](https://en.wikipedia.org/wiki/George_Karl)

- [San Antonio Spurs Roster, 1974](https://www.statscrew.com/basketball/roster/t-SAA/y-1974)

In [14]:
# replace the NaN value under 'weight' with Karl's playing weight, according to source
player_data_df["weight"] = player_data_df["weight"].fillna(185)

In [15]:
# check to make sure that row's weight data matches
print(player_data_df.loc[[2142]])

             name  year_start  year_end position height  weight    birth_date  \
2142  George Karl        1974      1978      NaN    NaN   185.0  May 12, 1952   

                           college  
2142  University of North Carolina  


**Checks out.  But since we're here, let's go ahead and adjust his position and height NaN values as well**

In [16]:
# according to the Spurs roster (cited above), George Karl was 6'2" 
player_data_df["height"] = player_data_df["height"].fillna("6-2")

In [17]:
# using the same Spurs roster, Karl's position is listed as guard (G)
player_data_df["position"] = player_data_df["position"].fillna("G")

**Checking to see that George Karl's row is now complete:**

In [18]:
print(player_data_df.loc[[2142]])

             name  year_start  year_end position height  weight    birth_date  \
2142  George Karl        1974      1978        G    6-2   185.0  May 12, 1952   

                           college  
2142  University of North Carolina  


In [19]:
null_counts(player_data_df)

name            0
year_start      0
year_end        0
position        0
height          0
weight          0
birth_date     31
college       302
dtype: int64

**^^Checks out.  Now 'height' has to be converted from a string object to an integer**

- Process:

>From the original 'height' column, isolate both feet and inches
        - recall that since the whole column is a string, '6-10' is has an index of 3, so if we're going to isolate the inches, we have to use ':4' as the index range.

>convert the feet and inches measurements into a new column called 'total_inches' 
>Convert 'foot-inch' measurements to inches

In [20]:
# isolate 'feet' measurements into new column

player_data_df["feet"] = player_data_df.height.str[:1] 

In [21]:
# isolate inches into new column

player_data_df["inches"] = player_data_df.height.str[2:4]

In [22]:
# check to see if the columns have been added

player_data_df

name  year_start  year_end position height  weight  \
0          Alaa Abdelnaby        1991      1995      F-C   6-10   240.0   
1         Zaid Abdul-Aziz        1969      1978      C-F    6-9   235.0   
2     Kareem Abdul-Jabbar        1970      1989        C    7-2   225.0   
3      Mahmoud Abdul-Rauf        1991      2001        G    6-1   162.0   
4       Tariq Abdul-Wahad        1998      2003        F    6-6   223.0   
...                   ...         ...       ...      ...    ...     ...   
4545           Ante Zizic        2018      2018      F-C   6-11   250.0   
4546             Jim Zoet        1983      1983        C    7-1   240.0   
4547            Bill Zopf        1971      1971        G    6-1   170.0   
4548          Ivica Zubac        2017      2018        C    7-1   265.0   
4549           Matt Zunic        1949      1949      G-F    6-3   195.0   

             birth_date                                college feet inches  
0         June 24, 1968                        Duke University    6     10  
1         April 7, 1946                  Iowa State University    6      9  
2        April 16, 1947  University of California, Los Angeles    7      2  
3         March 9, 1969             Louisiana State University    6      1  
4      November 3, 1974              San Jose State University    6      6  
...                 ...                                    ...  ...    ...  
4545    January 4, 1997                                    NaN    6     11  
4546  December 20, 1953                  Kent State University    7      1  
4547       June 7, 1948                    Duquesne University    6      1  
4548     March 18, 1997                                    NaN    7      1  
4549  December 19, 1919           George Washington University    6      3  

[4550 rows x 10 columns]

In [23]:
# check to see the datatypes of the two new columns
player_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4550 entries, 0 to 4549
Data columns (total 10 columns):
name          4550 non-null object
year_start    4550 non-null int64
year_end      4550 non-null int64
position      4550 non-null object
height        4550 non-null object
weight        4550 non-null float64
birth_date    4519 non-null object
college       4248 non-null object
feet          4550 non-null object
inches        4550 non-null object
dtypes: float64(1), int64(2), object(7)
memory usage: 355.6+ KB


In [24]:
# convert the 'feet' column to integer for later math 

player_data_df["feet"] = player_data_df["feet"].astype(int)

In [25]:
# convert the 'inches' column to integer for later math

player_data_df["inches"] = player_data_df["inches"].astype(int)

In [26]:
# convert feet to inches 

player_data_df["feet_in_inches"] = (player_data_df["feet"] * 12).to_frame("feet")

player_data_df

name  year_start  year_end position height  weight  \
0          Alaa Abdelnaby        1991      1995      F-C   6-10   240.0   
1         Zaid Abdul-Aziz        1969      1978      C-F    6-9   235.0   
2     Kareem Abdul-Jabbar        1970      1989        C    7-2   225.0   
3      Mahmoud Abdul-Rauf        1991      2001        G    6-1   162.0   
4       Tariq Abdul-Wahad        1998      2003        F    6-6   223.0   
...                   ...         ...       ...      ...    ...     ...   
4545           Ante Zizic        2018      2018      F-C   6-11   250.0   
4546             Jim Zoet        1983      1983        C    7-1   240.0   
4547            Bill Zopf        1971      1971        G    6-1   170.0   
4548          Ivica Zubac        2017      2018        C    7-1   265.0   
4549           Matt Zunic        1949      1949      G-F    6-3   195.0   

             birth_date                                college  feet  inches  \
0         June 24, 1968                        Duke University     6      10   
1         April 7, 1946                  Iowa State University     6       9   
2        April 16, 1947  University of California, Los Angeles     7       2   
3         March 9, 1969             Louisiana State University     6       1   
4      November 3, 1974              San Jose State University     6       6   
...                 ...                                    ...   ...     ...   
4545    January 4, 1997                                    NaN     6      11   
4546  December 20, 1953                  Kent State University     7       1   
4547       June 7, 1948                    Duquesne University     6       1   
4548     March 18, 1997                                    NaN     7       1   
4549  December 19, 1919           George Washington University     6       3   

      feet_in_inches  
0                 72  
1                 72  
2                 84  
3                 72  
4                 72  
...              ...  
4545              72  
4546              84  
4547              72  
4548              84  
4549              72  

[4550 rows x 11 columns]

In [27]:
# new column adding 'feet_in_inches' and 'inches' columns

player_data_df["total_inches"] = (player_data_df["feet_in_inches"] + player_data_df["inches"])

player_data_df

name  year_start  year_end position height  weight  \
0          Alaa Abdelnaby        1991      1995      F-C   6-10   240.0   
1         Zaid Abdul-Aziz        1969      1978      C-F    6-9   235.0   
2     Kareem Abdul-Jabbar        1970      1989        C    7-2   225.0   
3      Mahmoud Abdul-Rauf        1991      2001        G    6-1   162.0   
4       Tariq Abdul-Wahad        1998      2003        F    6-6   223.0   
...                   ...         ...       ...      ...    ...     ...   
4545           Ante Zizic        2018      2018      F-C   6-11   250.0   
4546             Jim Zoet        1983      1983        C    7-1   240.0   
4547            Bill Zopf        1971      1971        G    6-1   170.0   
4548          Ivica Zubac        2017      2018        C    7-1   265.0   
4549           Matt Zunic        1949      1949      G-F    6-3   195.0   

             birth_date                                college  feet  inches  \
0         June 24, 1968                        Duke University     6      10   
1         April 7, 1946                  Iowa State University     6       9   
2        April 16, 1947  University of California, Los Angeles     7       2   
3         March 9, 1969             Louisiana State University     6       1   
4      November 3, 1974              San Jose State University     6       6   
...                 ...                                    ...   ...     ...   
4545    January 4, 1997                                    NaN     6      11   
4546  December 20, 1953                  Kent State University     7       1   
4547       June 7, 1948                    Duquesne University     6       1   
4548     March 18, 1997                                    NaN     7       1   
4549  December 19, 1919           George Washington University     6       3   

      feet_in_inches  total_inches  
0                 72            82  
1                 72            81  
2                 84            86  
3                 72            73  
4                 72            78  
...              ...           ...  
4545              72            83  
4546              84            85  
4547              72            73  
4548              84            85  
4549              72            75  

[4550 rows x 12 columns]

**Now that we have the height of each player in total inches, we can replace the values in 'height' with the values in 'total inches'**

In [28]:
player_data_df = player_data_df.assign(height=player_data_df["total_inches"])

player_data_df

name  year_start  year_end position  height  weight  \
0          Alaa Abdelnaby        1991      1995      F-C      82   240.0   
1         Zaid Abdul-Aziz        1969      1978      C-F      81   235.0   
2     Kareem Abdul-Jabbar        1970      1989        C      86   225.0   
3      Mahmoud Abdul-Rauf        1991      2001        G      73   162.0   
4       Tariq Abdul-Wahad        1998      2003        F      78   223.0   
...                   ...         ...       ...      ...     ...     ...   
4545           Ante Zizic        2018      2018      F-C      83   250.0   
4546             Jim Zoet        1983      1983        C      85   240.0   
4547            Bill Zopf        1971      1971        G      73   170.0   
4548          Ivica Zubac        2017      2018        C      85   265.0   
4549           Matt Zunic        1949      1949      G-F      75   195.0   

             birth_date                                college  feet  inches  \
0         June 24, 1968                        Duke University     6      10   
1         April 7, 1946                  Iowa State University     6       9   
2        April 16, 1947  University of California, Los Angeles     7       2   
3         March 9, 1969             Louisiana State University     6       1   
4      November 3, 1974              San Jose State University     6       6   
...                 ...                                    ...   ...     ...   
4545    January 4, 1997                                    NaN     6      11   
4546  December 20, 1953                  Kent State University     7       1   
4547       June 7, 1948                    Duquesne University     6       1   
4548     March 18, 1997                                    NaN     7       1   
4549  December 19, 1919           George Washington University     6       3   

      feet_in_inches  total_inches  
0                 72            82  
1                 72            81  
2                 84            86  
3                 72            73  
4                 72            78  
...              ...           ...  
4545              72            83  
4546              84            85  
4547              72            73  
4548              84            85  
4549              72            75  

[4550 rows x 12 columns]

**To clean things up, we can now drop the columns 'feet', 'inches', 'feet_in_inches', and 'total_inches'.  After that, rename 'height' column to 'height_inches'** 

In [29]:
# drop the columns

player_data_df = player_data_df.drop(["feet", "inches", "feet_in_inches", 
                                      "total_inches"], axis=1)

In [30]:
# rename 'height' column

player_data_df = player_data_df.rename(columns={"height" : "height_inches"})

In [31]:
player_data_df

name  year_start  year_end position  height_inches  \
0          Alaa Abdelnaby        1991      1995      F-C             82   
1         Zaid Abdul-Aziz        1969      1978      C-F             81   
2     Kareem Abdul-Jabbar        1970      1989        C             86   
3      Mahmoud Abdul-Rauf        1991      2001        G             73   
4       Tariq Abdul-Wahad        1998      2003        F             78   
...                   ...         ...       ...      ...            ...   
4545           Ante Zizic        2018      2018      F-C             83   
4546             Jim Zoet        1983      1983        C             85   
4547            Bill Zopf        1971      1971        G             73   
4548          Ivica Zubac        2017      2018        C             85   
4549           Matt Zunic        1949      1949      G-F             75   

      weight         birth_date                                college  
0      240.0      June 24, 1968                        Duke University  
1      235.0      April 7, 1946                  Iowa State University  
2      225.0     April 16, 1947  University of California, Los Angeles  
3      162.0      March 9, 1969             Louisiana State University  
4      223.0   November 3, 1974              San Jose State University  
...      ...                ...                                    ...  
4545   250.0    January 4, 1997                                    NaN  
4546   240.0  December 20, 1953                  Kent State University  
4547   170.0       June 7, 1948                    Duquesne University  
4548   265.0     March 18, 1997                                    NaN  
4549   195.0  December 19, 1919           George Washington University  

[4550 rows x 8 columns]

In [32]:
player_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4550 entries, 0 to 4549
Data columns (total 8 columns):
name             4550 non-null object
year_start       4550 non-null int64
year_end         4550 non-null int64
position         4550 non-null object
height_inches    4550 non-null int64
weight           4550 non-null float64
birth_date       4519 non-null object
college          4248 non-null object
dtypes: float64(1), int64(3), object(4)
memory usage: 284.5+ KB


**Success!  The original heights have not only been converted from stuff like '6-10' to inches, but also the datatype of that column is now int64, and we can start the analysis.**

- Now to see what to do about the NaNs in 'birth_date' and 'college'

In [81]:
player_data_df[player_data_df["college"].isnull()].index

Int64Index([], dtype='int64')

From the list above, I'm going to pick five at random and look at their data:

In [34]:
print(player_data_df.iloc[[9, 32, 363, 285, 596]])

                  name  year_start  year_end position  height_inches  weight  \
9         Alex Abrines        2017      2018      G-F             78   190.0   
32       Alexis Ajinca        2009      2017        C             86   248.0   
363  Bogdan Bogdanovic        2018      2018        G             78   205.0   
285    Jonathan Bender        2000      2010        F             83   202.0   
596      Jose Calderon        2006      2018        G             75   200.0   

             birth_date college  
9        August 1, 1993     NaN  
32          May 6, 1988     NaN  
363     August 18, 1992     NaN  
285    January 30, 1981     NaN  
596  September 28, 1981     NaN  


In [35]:
print("Don Bielke" in player_data_df.name.values)
print("Don Bielke" in player_data_df.name.values)
print("Don Bielke" in player_data_df.name.values)
print("Don Bielke" in player_data_df.name.values)
print("Don Bielke" in player_data_df.name.values)

True
True
True
True
True


Googling these players has given me the following results:
    
- Alex Abrinas (index 9) played for Spain's National Team (Barca) with the Liga Endesa, which is basically Spain's NBA.
    
- Alexis Ajinca (index 32) came out of France's INSEP - the National Institute of Sport, Expertise, and Performance.  It's a place in France that trains elite athletes.
    
- Bogdan Bogdanovic (index 363) had a colorful past playing for Serbia's national team and Euro League. 
    
- Jonathan Bender (index 285) was selected in the NBA Draft out of High School (Picayune, MS).
    
- Jose Calderon (index 586) played with the Spanish National Team and Liga Endesa.

The NBA has pulled a ton of guys out of High School and the European Leagues, so it makes sense that the 'college' columns have NaN values because those players didn't go to college.

Best thing to do is to replace those NaNs in the 'college' column with 'No college listed':

In [36]:
player_data_df["college"] = player_data_df["college"].fillna("No college listed")

In [37]:
# check to see if the 'NaN' and '0' indexes match

print(player_data_df.iloc[[9, 32, 363, 285, 596]])

                  name  year_start  year_end position  height_inches  weight  \
9         Alex Abrines        2017      2018      G-F             78   190.0   
32       Alexis Ajinca        2009      2017        C             86   248.0   
363  Bogdan Bogdanovic        2018      2018        G             78   205.0   
285    Jonathan Bender        2000      2010        F             83   202.0   
596      Jose Calderon        2006      2018        G             75   200.0   

             birth_date            college  
9        August 1, 1993  No college listed  
32          May 6, 1988  No college listed  
363     August 18, 1992  No college listed  
285    January 30, 1981  No college listed  
596  September 28, 1981  No college listed  


**Checks out...**

- Now to do the same thing with the NaNs in the 'birth_date' column

In [38]:
player_data_df[player_data_df["birth_date"].isnull()].index

Int64Index([  49,  317,  473,  585,  800,  896, 1011, 1074, 1419, 1631, 2219,
            2360, 2470, 2598, 2608, 2817, 2821, 3016, 3092, 3110, 3336, 3569,
            3771, 3821, 3829, 3870, 3993, 4084, 4289, 4430, 4438],
           dtype='int64')

**Taking a look at who these people are:**

In [39]:
print(player_data_df.iloc[[49, 317, 1074, 2817, 3993]])

               name  year_start  year_end position  height_inches  weight  \
49       Bill Allen        1968      1968      C-F             80   205.0   
317      Don Bielke        1956      1956        C             79   240.0   
1074     Rich Dumas        1969      1969        G             75   170.0   
2817    Larry Moore        1968      1968        F             79   215.0   
3993  Willis Thomas        1968      1968        G             74   185.0   

     birth_date                        college  
49          NaN    New Mexico State University  
317         NaN          Valparaiso University  
1074        NaN  Northeastern State University  
2817        NaN              No college listed  
3993        NaN     Tennessee State University  


In [40]:
# replace 'birth_date' NaNs with 0

player_data_df["birth_date"] = player_data_df["birth_date"].fillna("Not listed")

In [41]:
# check

print(player_data_df.iloc[[49, 317, 1074, 2817, 3993]])

               name  year_start  year_end position  height_inches  weight  \
49       Bill Allen        1968      1968      C-F             80   205.0   
317      Don Bielke        1956      1956        C             79   240.0   
1074     Rich Dumas        1969      1969        G             75   170.0   
2817    Larry Moore        1968      1968        F             79   215.0   
3993  Willis Thomas        1968      1968        G             74   185.0   

      birth_date                        college  
49    Not listed    New Mexico State University  
317   Not listed          Valparaiso University  
1074  Not listed  Northeastern State University  
2817  Not listed              No college listed  
3993  Not listed     Tennessee State University  


In [42]:
# Take a final look at our cleaning efforts:

player_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4550 entries, 0 to 4549
Data columns (total 8 columns):
name             4550 non-null object
year_start       4550 non-null int64
year_end         4550 non-null int64
position         4550 non-null object
height_inches    4550 non-null int64
weight           4550 non-null float64
birth_date       4550 non-null object
college          4550 non-null object
dtypes: float64(1), int64(3), object(4)
memory usage: 284.5+ KB


## Players DataFrame Work

In [43]:
# bring in the 'player' csv

players_df = pd.read_csv("/Users/DataScience/Rimshotz/alley-oop-nba-stats/Players.csv")

b = players_df.isnull().sum().sum()

players_shape = players_df.shape
print("Player Data Information")
print(f"Consisting of {players_shape[0]} rows and {players_shape[1]} columns")
print(f"Not all data is present.  We still have {b} missing values.")
players_df

Player Data Information
Consisting of 3922 rows and 8 columns
Not all data is present.  We still have 1306 missing values.


Unnamed: 0             Player  height  weight  \
0              0    Curly Armstrong   180.0    77.0   
1              1       Cliff Barker   188.0    83.0   
2              2      Leo Barnhorst   193.0    86.0   
3              3         Ed Bartels   196.0    88.0   
4              4        Ralph Beard   178.0    79.0   
...          ...                ...     ...     ...   
3917        3917      Troy Williams   198.0    97.0   
3918        3918       Kyle Wiltjer   208.0   108.0   
3919        3919  Stephen Zimmerman   213.0   108.0   
3920        3920        Paul Zipser   203.0    97.0   
3921        3921        Ivica Zubac   216.0   120.0   

                              collage    born      birth_city  \
0                  Indiana University  1918.0             NaN   
1              University of Kentucky  1921.0        Yorktown   
2            University of Notre Dame  1924.0             NaN   
3     North Carolina State University  1925.0             NaN   
4              University of Kentucky  1927.0     Hardinsburg   
...                               ...     ...             ...   
3917  South Carolina State University  1969.0        Columbia   
3918               Gonzaga University  1992.0        Portland   
3919  University of Nevada, Las Vegas  1996.0  Hendersonville   
3920                              NaN  1994.0      Heidelberg   
3921                              NaN  1997.0          Mostar   

                 birth_state  
0                        NaN  
1                    Indiana  
2                        NaN  
3                        NaN  
4                   Kentucky  
...                      ...  
3917          South Carolina  
3918                  Oregon  
3919               Tennessee  
3920                 Germany  
3921  Bosnia and Herzegovina  

[3922 rows x 8 columns]

In [44]:
players_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3922 entries, 0 to 3921
Data columns (total 8 columns):
Unnamed: 0     3922 non-null int64
Player         3921 non-null object
height         3921 non-null float64
weight         3921 non-null float64
collage        3573 non-null object
born           3921 non-null float64
birth_city     3452 non-null object
birth_state    3439 non-null object
dtypes: float64(3), int64(1), object(4)
memory usage: 245.2+ KB


**Right off the bat, there's a misspelling: 'collage' should be 'college'...**

In [45]:
# renaming "collage" column to "college" - nobody plays basketball in a collage...

players_df = players_df.rename(columns={"collage" : "college"})

players_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3922 entries, 0 to 3921
Data columns (total 8 columns):
Unnamed: 0     3922 non-null int64
Player         3921 non-null object
height         3921 non-null float64
weight         3921 non-null float64
college        3573 non-null object
born           3921 non-null float64
birth_city     3452 non-null object
birth_state    3439 non-null object
dtypes: float64(3), int64(1), object(4)
memory usage: 245.2+ KB


**'Unnamed' is just a column that repeats the index, so we can drop it:**

In [46]:
players_df = players_df.drop(columns="Unnamed: 0", axis=1)

players_df.head()

Player  height  weight                          college    born  \
0  Curly Armstrong   180.0    77.0               Indiana University  1918.0   
1     Cliff Barker   188.0    83.0           University of Kentucky  1921.0   
2    Leo Barnhorst   193.0    86.0         University of Notre Dame  1924.0   
3       Ed Bartels   196.0    88.0  North Carolina State University  1925.0   
4      Ralph Beard   178.0    79.0           University of Kentucky  1927.0   

    birth_city birth_state  
0          NaN         NaN  
1     Yorktown     Indiana  
2          NaN         NaN  
3          NaN         NaN  
4  Hardinsburg    Kentucky

In [47]:
players_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3922 entries, 0 to 3921
Data columns (total 7 columns):
Player         3921 non-null object
height         3921 non-null float64
weight         3921 non-null float64
college        3573 non-null object
born           3921 non-null float64
birth_city     3452 non-null object
birth_state    3439 non-null object
dtypes: float64(3), object(4)
memory usage: 214.6+ KB


In [48]:
null_feedback(players_df)

Total number of NaNs by column: 
Player           1
height           1
weight           1
college        349
born             1
birth_city     470
birth_state    483
dtype: int64.
Percentage of columns that are NaN: 
(Player          0.0
height          0.0
weight          0.0
college         9.0
born            0.0
birth_city     12.0
birth_state    12.0
dtype: float64, 2).


**Overview of this dataset reveals the following:**

1.) there's a NaN in 'Player' - can we just drop the whole row?
    
2.) to stay consistent with 'player_data_df' info, we'll need to convert the 'height' column from centimeters to inches

3.) we will have to convert 'weight' from kilogram to pounds to keep consistent with 'player_data_df'

4.) is it just me, or is that '.0' after the 'born' year seem offensive.  "English, people!!"

In [49]:
#1.) check the row where 'Player' data is 'NaN'

players_df[players_df["Player"].isnull()].index.tolist()

[223]

In [50]:
# take a look at row 223

players_df.loc[[223]]

Player  height  weight college  born birth_city birth_state
223    NaN     NaN     NaN     NaN   NaN        NaN         NaN

In [51]:
# yup.  We can drop that entire row

players_df = players_df.drop([223])

players_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3921 entries, 0 to 3921
Data columns (total 7 columns):
Player         3921 non-null object
height         3921 non-null float64
weight         3921 non-null float64
college        3573 non-null object
born           3921 non-null float64
birth_city     3452 non-null object
birth_state    3439 non-null object
dtypes: float64(3), object(4)
memory usage: 245.1+ KB


In [52]:
#2.) Convert 'height' from centimeters to inches - each inch = 2.54 centimeters
players_df["height"] = players_df.height.div(2.54).round()

# get rid of decimal in height
players_df["height"] = players_df["height"].astype(int)

players_df.head()

Player  height  weight                          college    born  \
0  Curly Armstrong      71    77.0               Indiana University  1918.0   
1     Cliff Barker      74    83.0           University of Kentucky  1921.0   
2    Leo Barnhorst      76    86.0         University of Notre Dame  1924.0   
3       Ed Bartels      77    88.0  North Carolina State University  1925.0   
4      Ralph Beard      70    79.0           University of Kentucky  1927.0   

    birth_city birth_state  
0          NaN         NaN  
1     Yorktown     Indiana  
2          NaN         NaN  
3          NaN         NaN  
4  Hardinsburg    Kentucky

In [53]:
#3.) Convert 'weight' from kilograms to pounds - each kilo = 2.2 pounds

players_df["weight"] = players_df.weight.mul(2.2).round()

# get rid of decimal point

players_df["weight"] = players_df["weight"].astype(int)

players_df.head()

Player  height  weight                          college    born  \
0  Curly Armstrong      71     169               Indiana University  1918.0   
1     Cliff Barker      74     183           University of Kentucky  1921.0   
2    Leo Barnhorst      76     189         University of Notre Dame  1924.0   
3       Ed Bartels      77     194  North Carolina State University  1925.0   
4      Ralph Beard      70     174           University of Kentucky  1927.0   

    birth_city birth_state  
0          NaN         NaN  
1     Yorktown     Indiana  
2          NaN         NaN  
3          NaN         NaN  
4  Hardinsburg    Kentucky

In [54]:
#4.) getting rid of that '.0' cr*p in 'born'

players_df["born"] = players_df["born"].astype(int)

players_df.head()

Player  height  weight                          college  born  \
0  Curly Armstrong      71     169               Indiana University  1918   
1     Cliff Barker      74     183           University of Kentucky  1921   
2    Leo Barnhorst      76     189         University of Notre Dame  1924   
3       Ed Bartels      77     194  North Carolina State University  1925   
4      Ralph Beard      70     174           University of Kentucky  1927   

    birth_city birth_state  
0          NaN         NaN  
1     Yorktown     Indiana  
2          NaN         NaN  
3          NaN         NaN  
4  Hardinsburg    Kentucky

**Addressing nulls in the dataset:**

As in the previous dataframe (player_data_df), there are a substantial amount of NaN values throughout the data, namely **'college (9%),' 'birth_city (12%),'** and **'birth_state (12%).'**  That's a lot of data, and to go through and look up each individual case would be insanely time consuming.  Rather, we will replace 'NaN' with 'Not Available' until we start doing the analysis.  For those names that come up where there are 'Not Available' values, we'll fix them on a case-by-case basis. 
    

In [55]:
# replacing nans in columns with 'not available'

columns = ["college", "birth_city", "birth_state"]

players_df[columns] = players_df[columns].fillna("Not Available")

players_df.head()

Player  height  weight                          college  born  \
0  Curly Armstrong      71     169               Indiana University  1918   
1     Cliff Barker      74     183           University of Kentucky  1921   
2    Leo Barnhorst      76     189         University of Notre Dame  1924   
3       Ed Bartels      77     194  North Carolina State University  1925   
4      Ralph Beard      70     174           University of Kentucky  1927   

      birth_city    birth_state  
0  Not Available  Not Available  
1       Yorktown        Indiana  
2  Not Available  Not Available  
3  Not Available  Not Available  
4    Hardinsburg       Kentucky

In [56]:
players_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3921 entries, 0 to 3921
Data columns (total 7 columns):
Player         3921 non-null object
height         3921 non-null int64
weight         3921 non-null int64
college        3921 non-null object
born           3921 non-null int64
birth_city     3921 non-null object
birth_state    3921 non-null object
dtypes: int64(3), object(4)
memory usage: 245.1+ KB


**No more null values, strings and ints are where they're supposed to be.  Done.  On to the next dataframe...**

## Seasons_Stats DataFrame Work

In [57]:
# bring in the 'seasons_stats' csv

seasons_stats_df = pd.read_csv("/Users/DataScience/Rimshotz/alley-oop-nba-stats/Seasons_Stats.csv")

c = seasons_stats_df.isnull().sum().sum()

seasons_stats_shape = seasons_stats_df.shape
print("Statistical Information by Season")
print(f"Consisting of {seasons_stats_shape[0]} rows and {seasons_stats_shape[1]} columns")
print(f"While informative, still needs work: we are missing {c} values.")
seasons_stats_df

Statistical Information by Season
Consisting of 24691 rows and 53 columns
While informative, still needs work: we are missing 154919 values.


Unnamed: 0    Year             Player  Pos   Age   Tm     G    GS  \
0               0  1950.0    Curly Armstrong  G-F  31.0  FTW  63.0   NaN   
1               1  1950.0       Cliff Barker   SG  29.0  INO  49.0   NaN   
2               2  1950.0      Leo Barnhorst   SF  25.0  CHS  67.0   NaN   
3               3  1950.0         Ed Bartels    F  24.0  TOT  15.0   NaN   
4               4  1950.0         Ed Bartels    F  24.0  DNN  13.0   NaN   
...           ...     ...                ...  ...   ...  ...   ...   ...   
24686       24686  2017.0        Cody Zeller   PF  24.0  CHO  62.0  58.0   
24687       24687  2017.0       Tyler Zeller    C  27.0  BOS  51.0   5.0   
24688       24688  2017.0  Stephen Zimmerman    C  20.0  ORL  19.0   0.0   
24689       24689  2017.0        Paul Zipser   SF  22.0  CHI  44.0  18.0   
24690       24690  2017.0        Ivica Zubac    C  19.0  LAL  38.0  11.0   

           MP   PER  ...    FT%    ORB    DRB    TRB    AST   STL   BLK   TOV  \
0         NaN   NaN  ...  0.705    NaN    NaN    NaN  176.0   NaN   NaN   NaN   
1         NaN   NaN  ...  0.708    NaN    NaN    NaN  109.0   NaN   NaN   NaN   
2         NaN   NaN  ...  0.698    NaN    NaN    NaN  140.0   NaN   NaN   NaN   
3         NaN   NaN  ...  0.559    NaN    NaN    NaN   20.0   NaN   NaN   NaN   
4         NaN   NaN  ...  0.548    NaN    NaN    NaN   20.0   NaN   NaN   NaN   
...       ...   ...  ...    ...    ...    ...    ...    ...   ...   ...   ...   
24686  1725.0  16.7  ...  0.679  135.0  270.0  405.0   99.0  62.0  58.0  65.0   
24687   525.0  13.0  ...  0.564   43.0   81.0  124.0   42.0   7.0  21.0  20.0   
24688   108.0   7.3  ...  0.600   11.0   24.0   35.0    4.0   2.0   5.0   3.0   
24689   843.0   6.9  ...  0.775   15.0  110.0  125.0   36.0  15.0  16.0  40.0   
24690   609.0  17.0  ...  0.653   41.0  118.0  159.0   30.0  14.0  33.0  30.0   

          PF    PTS  
0      217.0  458.0  
1       99.0  279.0  
2      192.0  438.0  
3       29.0   63.0  
4       27.0   59.0  
...      ...    ...  
24686  189.0  639.0  
24687   61.0  178.0  
24688   17.0   23.0  
24689   78.0  240.0  
24690   66.0  284.0  

[24691 rows x 53 columns]

In [58]:
seasons_stats_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24691 entries, 0 to 24690
Data columns (total 53 columns):
Unnamed: 0    24691 non-null int64
Year          24624 non-null float64
Player        24624 non-null object
Pos           24624 non-null object
Age           24616 non-null float64
Tm            24624 non-null object
G             24624 non-null float64
GS            18233 non-null float64
MP            24138 non-null float64
PER           24101 non-null float64
TS%           24538 non-null float64
3PAr          18839 non-null float64
FTr           24525 non-null float64
ORB%          20792 non-null float64
DRB%          20792 non-null float64
TRB%          21571 non-null float64
AST%          22555 non-null float64
STL%          20792 non-null float64
BLK%          20792 non-null float64
TOV%          19582 non-null float64
USG%          19640 non-null float64
blanl         0 non-null float64
OWS           24585 non-null float64
DWS           24585 non-null float64
WS          

**Initial observations from first glance at seasons_stats_df:**

1.) Loads of abbreviations on the column names.  Going to have to rename the columns for clarity;

2.) Rows 'blanl' and 'blank2' are complete rubbish, so we can drop them;

3.) 'Unnamed: 0' column is just a repeat of the index, so we can drop it;

4.) A lot of columns are floats - we need to change it to an integer types. For example, 'GS,' or 'Games Started'.  No need for a float value because you either start a game, or you don't;  

5.) Looks like stats have changed over time.  Lots of null values for the earlier data - may need to divide this data into two separate dataframes;

6.) This will need a data dictionary, as all columns are abbreviations.  Since we're doing that (finding out what the abbreviations are), it may be a good idea to rename those columns with what we find

In [59]:
# rename columns for sake of clarity
seasons_stats_df.rename(
        columns={
            "Year": "year",
            "Player": "player",
            "Pos": "position",
            "Age": "age",
            "Tm": "team",
            "G": "games",
            "GS": "games_started",
            "MP": "minutes_played",
            "PER": "player_efficiency",
            "TS%": "true_shooting_%",
            "3PAr": "three_pt_tries",
            "FTr": "free_throws",
            "ORB%": "off_rebound_%",
            "DRB%": "def_rebound_%",
            "TRB%": "total_rebound_%",
            "AST%": "assist_%",
            "STL%": "steal_%",
            "BLK%": "block_%",
            "TOV%" : "turnover_%",
            "USG%": "usage_%",
            "blanl": "blank1",
            "OWS": "offensive_win_shares",
            "DWS": "defensive_win_shares",
            "WS": "win_shares",
            "WS/48": "win_shares_per_48min",
            "OBPM": "off_box_plus_minus",
            "DBPM": "def_box_plus_minus",
            "BPM": "box_plus_minus",
            "VORP": "value_over_replacement",
            "FG" : "field_goals", 
            "FGA" : "field_goal_attempts",
            "FG%" : "field_goal_%", 
            "3P" : "3_pointers", 
            "3PA" : "3_point_tries",
            "3P%" : "3_point_%",
            "2P" : "2_pointers",
            "2PA" : "2_point_tries",
            "2P%" : "2_point_%",
            "eFG%" : "effective_field_goal_%",
            "FT" : "free_throws",
            "FTA" : "free_throw_attempts",
            "FT%" : "free_throw_%",
            "ORB" : "off_rebounds",
            "DRB" : "def_rebounds",
            "TRB" : "total_rebounds",
            "AST" : "assists",
            "STL" : "steals",
            "BLK" : "blocks",
            "TOV" : "turnovers",
            "PF" : "personal_fouls",
            "PTS" : "points",
        },
        inplace=True,
        )

In [60]:
seasons_stats_df.head()

Unnamed: 0    year           player position   age team  games  \
0           0  1950.0  Curly Armstrong      G-F  31.0  FTW   63.0   
1           1  1950.0     Cliff Barker       SG  29.0  INO   49.0   
2           2  1950.0    Leo Barnhorst       SF  25.0  CHS   67.0   
3           3  1950.0       Ed Bartels        F  24.0  TOT   15.0   
4           4  1950.0       Ed Bartels        F  24.0  DNN   13.0   

   games_started  minutes_played  player_efficiency  ...  free_throw_%  \
0            NaN             NaN                NaN  ...         0.705   
1            NaN             NaN                NaN  ...         0.708   
2            NaN             NaN                NaN  ...         0.698   
3            NaN             NaN                NaN  ...         0.559   
4            NaN             NaN                NaN  ...         0.548   

   off_rebounds  def_rebounds  total_rebounds  assists  steals  blocks  \
0           NaN           NaN             NaN    176.0     NaN     NaN   
1           NaN           NaN             NaN    109.0     NaN     NaN   
2           NaN           NaN             NaN    140.0     NaN     NaN   
3           NaN           NaN             NaN     20.0     NaN     NaN   
4           NaN           NaN             NaN     20.0     NaN     NaN   

   turnovers  personal_fouls  points  
0        NaN           217.0   458.0  
1        NaN            99.0   279.0  
2        NaN           192.0   438.0  
3        NaN            29.0    63.0  
4        NaN            27.0    59.0  

[5 rows x 53 columns]

**Based on what we observed earlier, the following columns can be dropped because they lack necessary data for exploration:**

| Column To Be Dropped | Reason            |
|:--------------------:|:-----------------:|
| "Unnamed: 0"         | replica of index  |
| "blank1"             | all values are NaN| 
| "blank2"             | all values are NaN| 

In [61]:
# drop 'blank1' and 'blank2' columns - 100% of the values are missing

seasons_stats_df = seasons_stats_df.drop(["Unnamed: 0","blank1", "blank2"], axis=1)
seasons_stats_df

year             player position   age team  games  games_started  \
0      1950.0    Curly Armstrong      G-F  31.0  FTW   63.0            NaN   
1      1950.0       Cliff Barker       SG  29.0  INO   49.0            NaN   
2      1950.0      Leo Barnhorst       SF  25.0  CHS   67.0            NaN   
3      1950.0         Ed Bartels        F  24.0  TOT   15.0            NaN   
4      1950.0         Ed Bartels        F  24.0  DNN   13.0            NaN   
...       ...                ...      ...   ...  ...    ...            ...   
24686  2017.0        Cody Zeller       PF  24.0  CHO   62.0           58.0   
24687  2017.0       Tyler Zeller        C  27.0  BOS   51.0            5.0   
24688  2017.0  Stephen Zimmerman        C  20.0  ORL   19.0            0.0   
24689  2017.0        Paul Zipser       SF  22.0  CHI   44.0           18.0   
24690  2017.0        Ivica Zubac        C  19.0  LAL   38.0           11.0   

       minutes_played  player_efficiency  true_shooting_%  ...  free_throw_%  \
0                 NaN                NaN            0.368  ...         0.705   
1                 NaN                NaN            0.435  ...         0.708   
2                 NaN                NaN            0.394  ...         0.698   
3                 NaN                NaN            0.312  ...         0.559   
4                 NaN                NaN            0.308  ...         0.548   
...               ...                ...              ...  ...           ...   
24686          1725.0               16.7            0.604  ...         0.679   
24687           525.0               13.0            0.508  ...         0.564   
24688           108.0                7.3            0.346  ...         0.600   
24689           843.0                6.9            0.503  ...         0.775   
24690           609.0               17.0            0.547  ...         0.653   

       off_rebounds  def_rebounds  total_rebounds  assists  steals  blocks  \
0               NaN           NaN             NaN    176.0     NaN     NaN   
1               NaN           NaN             NaN    109.0     NaN     NaN   
2               NaN           NaN             NaN    140.0     NaN     NaN   
3               NaN           NaN             NaN     20.0     NaN     NaN   
4               NaN           NaN             NaN     20.0     NaN     NaN   
...             ...           ...             ...      ...     ...     ...   
24686         135.0         270.0           405.0     99.0    62.0    58.0   
24687          43.0          81.0           124.0     42.0     7.0    21.0   
24688          11.0          24.0            35.0      4.0     2.0     5.0   
24689          15.0         110.0           125.0     36.0    15.0    16.0   
24690          41.0         118.0           159.0     30.0    14.0    33.0   

       turnovers  personal_fouls  points  
0            NaN           217.0   458.0  
1            NaN            99.0   279.0  
2            NaN           192.0   438.0  
3            NaN            29.0    63.0  
4            NaN            27.0    59.0  
...          ...             ...     ...  
24686       65.0           189.0   639.0  
24687       20.0            61.0   178.0  
24688        3.0            17.0    23.0  
24689       40.0            78.0   240.0  
24690       30.0            66.0   284.0  

[24691 rows x 50 columns]

In [62]:
# change dtypes from floats to integers in the appropriate columns

# cols = ["year", "age", "games", "assists", "personal_fouls", "points"]

# seasons_stats_df[cols] = seasons_stats_df[cols].astype(int)

# seasons_stats_df.head()

In [63]:
# Code above (In[66]) returned an error b/c of the NaN values.  
# Can't make them integers.  So need to check for NaNs in those columns and replace them

**Looking at "year", "age", "games", "assists", "personal_fouls", "points"**

In [64]:
print(f"There are {seasons_stats_df.year.isna().sum()} missing values in 'year' column.")
print(f"There are {seasons_stats_df.age.isna().sum()} values missing in 'age' column.")
print(f"'Games' column is missing {seasons_stats_df.games.isna().sum()} values.")
print(f"'Assists' is missing {seasons_stats_df.assists.isna().sum()} values.")
print(f"Apparently, {seasons_stats_df.personal_fouls.isna().sum()} personal fouls weren't recorded.")
print(f"Lastly, in {seasons_stats_df.points.isna().sum()} cases, nobody scored any points")

There are 67 missing values in 'year' column.
There are 75 values missing in 'age' column.
'Games' column is missing 67 values.
'Assists' is missing 67 values.
Apparently, 67 personal fouls weren't recorded.
Lastly, in 67 cases, nobody scored any points


**Lots of '67' counts in there(^^).  Time for a closer look...**

In [65]:
# taking a look at the indexes where these values are missing

seasons_stats_df[seasons_stats_df["year"].isnull()].index

Int64Index([  312,   487,   618,   779,   911,  1021,  1128,  1236,  1348,
             1459,  1577,  1682,  1808,  1942,  2078,  2211,  2347,  2481,
             2659,  2866,  3068,  3314,  3580,  3850,  4096,  4373,  4648,
             5006,  5381,  5726,  6084,  6448,  6822,  7214,  7558,  7921,
             8301,  8680,  9107,  9546, 10006, 10448, 10907, 11357, 11839,
            12292, 12838, 13413, 13961, 14469, 14966, 15504, 16005, 16489,
            17075, 17661, 18225, 18742, 19338, 19921, 20500, 21126, 21678,
            22252, 22864, 23516, 24095],
           dtype='int64')

In [66]:
# specific row (picking at random)

print(seasons_stats_df.iloc[[312, 2866, 4096, 6822, 11357]])

       year player position  age team  games  games_started  minutes_played  \
312     NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
2866    NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
4096    NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
6822    NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
11357   NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   

       player_efficiency  true_shooting_%  ...  free_throw_%  off_rebounds  \
312                  NaN              NaN  ...           NaN           NaN   
2866                 NaN              NaN  ...           NaN           NaN   
4096                 NaN              NaN  ...           NaN           NaN   
6822                 NaN              NaN  ...           NaN           NaN   
11357                NaN              NaN  ...           NaN           NaN   

       def_rebounds  total_rebounds  assists  steals  bl

**Huh.  Zero data.  What about 5 other rows (again: picked at random)?**

In [67]:
print(seasons_stats_df.iloc[[5381, 2659, 18225, 911, 22864]])

       year player position  age team  games  games_started  minutes_played  \
5381    NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
2659    NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
18225   NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
911     NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   
22864   NaN    NaN      NaN  NaN  NaN    NaN            NaN             NaN   

       player_efficiency  true_shooting_%  ...  free_throw_%  off_rebounds  \
5381                 NaN              NaN  ...           NaN           NaN   
2659                 NaN              NaN  ...           NaN           NaN   
18225                NaN              NaN  ...           NaN           NaN   
911                  NaN              NaN  ...           NaN           NaN   
22864                NaN              NaN  ...           NaN           NaN   

       def_rebounds  total_rebounds  assists  steals  bl

**After looking at this data, I think it's okay to drop these rows from our dataframe.**

In [68]:
# rows to drop come from the 'seasons_stats_df[seasons_stats_df["year"].isnull()].index'
# line above

# seasons_stats_df = seasons_stats_df.drop([312,   487,   618,   779,   911,  1021,  1128,  1236,  1348,
#              1459,  1577,  1682,  1808,  1942,  2078,  2211,  2347,  2481,
#              2659,  2866,  3068,  3314,  3580,  3850,  4096,  4373,  4648,
#              5006,  5381,  5726,  6084,  6448,  6822,  7214,  7558,  7921,
#              8301,  8680,  9107,  9546, 10006, 10448, 10907, 11357, 11839,
#             12292, 12838, 13413, 13961, 14469, 14966, 15504, 16005, 16489,
#             17075, 17661, 18225, 18742, 19338, 19921, 20500, 21126, 21678,
#             22252, 22864, 23516, 24095])

**But what about the data that was missing from the 'age' column?  The one that did NOT have a value of '67'?**

In [69]:
seasons_stats_df[seasons_stats_df["age"].isnull()].index

Int64Index([  312,   487,   618,   733,   779,   871,   911,   971,  1021,
             1028,  1104,  1105,  1106,  1128,  1214,  1236,  1348,  1459,
             1577,  1682,  1808,  1942,  2078,  2211,  2347,  2481,  2659,
             2866,  3068,  3314,  3580,  3850,  4096,  4373,  4648,  5006,
             5381,  5726,  6084,  6448,  6822,  7214,  7558,  7921,  8301,
             8680,  9107,  9546, 10006, 10448, 10907, 11357, 11839, 12292,
            12838, 13413, 13961, 14469, 14966, 15504, 16005, 16489, 17075,
            17661, 18225, 18742, 19338, 19921, 20500, 21126, 21678, 22252,
            22864, 23516, 24095],
           dtype='int64')

In [70]:
# taking a closer look

print(seasons_stats_df.iloc[[733, 871, 971, 1028, 1104, 1105, 1106, 1214]])

        year         player position  age team  games  games_started  \
733   1953.0   Mike O'Neill       SF  NaN  MLH    4.0            NaN   
871   1954.0  Frank Reddout        F  NaN  ROC    7.0            NaN   
971   1955.0    Ken McBride       SG  NaN  MLH   12.0            NaN   
1028  1956.0     Don Bielke        C  NaN  FTW    7.0            NaN   
1104  1956.0    Bob Schafer       SG  NaN  TOT   54.0            NaN   
1105  1956.0    Bob Schafer       SG  NaN  PHW   12.0            NaN   
1106  1956.0    Bob Schafer       SG  NaN  STL   42.0            NaN   
1214  1957.0    Bob Schafer       SG  NaN  SYR   11.0            NaN   

      minutes_played  player_efficiency  true_shooting_%  ...  free_throw_%  \
733             50.0                4.6            0.320  ...         1.000   
871             18.0               42.5            0.838  ...         0.750   
971            249.0                7.9            0.366  ...         0.724   
1028            38.0               

In [71]:
# checking to see if these names are in the player_data_df info

print("Mike O'Neill" in player_data_df.name.values)
print("Frank Reddout" in player_data_df.name.values)
print("Ken McBride" in player_data_df.name.values)
print("Don Bielke" in player_data_df.name.values)
print("Bob Schafer" in player_data_df.name.values)

True
True
True
True
True


**Given the size of this dataframe, it could be that we can replace the 'age' with 'born' data on the players_data_df.  Therefore we can zero them out here as long as we combine the dataframes later**

In [72]:
# replacing NaN ages with 1's
seasons_stats_df["age"] = seasons_stats_df["age"].fillna(1)

/usr/local/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [73]:
# check
print(seasons_stats_df.iloc[[733, 871, 971, 1028, 1104, 1105, 1106, 1214]])

        year         player position  age team  games  games_started  \
733   1953.0   Mike O'Neill       SF  1.0  MLH    4.0            NaN   
871   1954.0  Frank Reddout        F  1.0  ROC    7.0            NaN   
971   1955.0    Ken McBride       SG  1.0  MLH   12.0            NaN   
1028  1956.0     Don Bielke        C  1.0  FTW    7.0            NaN   
1104  1956.0    Bob Schafer       SG  1.0  TOT   54.0            NaN   
1105  1956.0    Bob Schafer       SG  1.0  PHW   12.0            NaN   
1106  1956.0    Bob Schafer       SG  1.0  STL   42.0            NaN   
1214  1957.0    Bob Schafer       SG  1.0  SYR   11.0            NaN   

      minutes_played  player_efficiency  true_shooting_%  ...  free_throw_%  \
733             50.0                4.6            0.320  ...         1.000   
871             18.0               42.5            0.838  ...         0.750   
971            249.0                7.9            0.366  ...         0.724   
1028            38.0               

**Done.  Now that we have seen a lot of the datatypes were were wanting to change were either useless or easily replacable, we can drop the rows with no values and change column datatypes from floats to integers**

In [74]:
# drop

seasons_stats_df = seasons_stats_df.drop([312,   487,   618,   779,   911,  1021,  1128,  1236,  1348,
             1459,  1577,  1682,  1808,  1942,  2078,  2211,  2347,  2481,
             2659,  2866,  3068,  3314,  3580,  3850,  4096,  4373,  4648,
             5006,  5381,  5726,  6084,  6448,  6822,  7214,  7558,  7921,
             8301,  8680,  9107,  9546, 10006, 10448, 10907, 11357, 11839,
            12292, 12838, 13413, 13961, 14469, 14966, 15504, 16005, 16489,
            17075, 17661, 18225, 18742, 19338, 19921, 20500, 21126, 21678,
            22252, 22864, 23516, 24095])

In [75]:
# change dtypes from floats to integers in the appropriate columns

cols = ["year", "age", "games", "assists", "personal_fouls", "points"]

seasons_stats_df[cols] = seasons_stats_df[cols].astype(int)

seasons_stats_df.head()

year           player position  age team  games  games_started  \
0  1950  Curly Armstrong      G-F   31  FTW     63            NaN   
1  1950     Cliff Barker       SG   29  INO     49            NaN   
2  1950    Leo Barnhorst       SF   25  CHS     67            NaN   
3  1950       Ed Bartels        F   24  TOT     15            NaN   
4  1950       Ed Bartels        F   24  DNN     13            NaN   

   minutes_played  player_efficiency  true_shooting_%  ...  free_throw_%  \
0             NaN                NaN            0.368  ...         0.705   
1             NaN                NaN            0.435  ...         0.708   
2             NaN                NaN            0.394  ...         0.698   
3             NaN                NaN            0.312  ...         0.559   
4             NaN                NaN            0.308  ...         0.548   

   off_rebounds  def_rebounds  total_rebounds  assists  steals  blocks  \
0           NaN           NaN             NaN      176     NaN     NaN   
1           NaN           NaN             NaN      109     NaN     NaN   
2           NaN           NaN             NaN      140     NaN     NaN   
3           NaN           NaN             NaN       20     NaN     NaN   
4           NaN           NaN             NaN       20     NaN     NaN   

   turnovers  personal_fouls  points  
0        NaN             217     458  
1        NaN              99     279  
2        NaN             192     438  
3        NaN              29      63  
4        NaN              27      59  

[5 rows x 50 columns]

In [76]:
# rechecking data

seasons_stats_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24624 entries, 0 to 24690
Data columns (total 50 columns):
year                      24624 non-null int64
player                    24624 non-null object
position                  24624 non-null object
age                       24624 non-null int64
team                      24624 non-null object
games                     24624 non-null int64
games_started             18233 non-null float64
minutes_played            24138 non-null float64
player_efficiency         24101 non-null float64
true_shooting_%           24538 non-null float64
three_pt_tries            18839 non-null float64
free_throws               24525 non-null float64
off_rebound_%             20792 non-null float64
def_rebound_%             20792 non-null float64
total_rebound_%           21571 non-null float64
assist_%                  22555 non-null float64
steal_%                   20792 non-null float64
block_%                   20792 non-null float64
turnover_%        

**Given the amount of nulls we're dealing with in columns like 'turnovers' and 'def_rebounds,' we can replace those missing values with '1.'  For instance: it would make sense that every player had *at least* 1 defensive rebound that year.**

- Also, for statistical modeling later on, '1' beats '0' every time.

In [77]:
seasons_stats_df = seasons_stats_df.fillna(1)

seasons_stats_df.head()

year           player position  age team  games  games_started  \
0  1950  Curly Armstrong      G-F   31  FTW     63            1.0   
1  1950     Cliff Barker       SG   29  INO     49            1.0   
2  1950    Leo Barnhorst       SF   25  CHS     67            1.0   
3  1950       Ed Bartels        F   24  TOT     15            1.0   
4  1950       Ed Bartels        F   24  DNN     13            1.0   

   minutes_played  player_efficiency  true_shooting_%  ...  free_throw_%  \
0             1.0                1.0            0.368  ...         0.705   
1             1.0                1.0            0.435  ...         0.708   
2             1.0                1.0            0.394  ...         0.698   
3             1.0                1.0            0.312  ...         0.559   
4             1.0                1.0            0.308  ...         0.548   

   off_rebounds  def_rebounds  total_rebounds  assists  steals  blocks  \
0           1.0           1.0             1.0      176     1.0     1.0   
1           1.0           1.0             1.0      109     1.0     1.0   
2           1.0           1.0             1.0      140     1.0     1.0   
3           1.0           1.0             1.0       20     1.0     1.0   
4           1.0           1.0             1.0       20     1.0     1.0   

   turnovers  personal_fouls  points  
0        1.0             217     458  
1        1.0              99     279  
2        1.0             192     438  
3        1.0              29      63  
4        1.0              27      59  

[5 rows x 50 columns]

**No sense in all these new '1' values to be floats, so changing them to integers.  Stuff like 'player_efficiency' though should remain a float**

In [78]:
more_cols = ["games_started", "minutes_played", "off_rebounds", "def_rebounds",
            "total_rebounds", "steals", "blocks", "turnovers"]

seasons_stats_df[more_cols] = seasons_stats_df[more_cols].astype(int)

seasons_stats_df.head()

year           player position  age team  games  games_started  \
0  1950  Curly Armstrong      G-F   31  FTW     63              1   
1  1950     Cliff Barker       SG   29  INO     49              1   
2  1950    Leo Barnhorst       SF   25  CHS     67              1   
3  1950       Ed Bartels        F   24  TOT     15              1   
4  1950       Ed Bartels        F   24  DNN     13              1   

   minutes_played  player_efficiency  true_shooting_%  ...  free_throw_%  \
0               1                1.0            0.368  ...         0.705   
1               1                1.0            0.435  ...         0.708   
2               1                1.0            0.394  ...         0.698   
3               1                1.0            0.312  ...         0.559   
4               1                1.0            0.308  ...         0.548   

   off_rebounds  def_rebounds  total_rebounds  assists  steals  blocks  \
0             1             1               1      176       1       1   
1             1             1               1      109       1       1   
2             1             1               1      140       1       1   
3             1             1               1       20       1       1   
4             1             1               1       20       1       1   

   turnovers  personal_fouls  points  
0          1             217     458  
1          1              99     279  
2          1             192     438  
3          1              29      63  
4          1              27      59  

[5 rows x 50 columns]

In [79]:
seasons_stats_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24624 entries, 0 to 24690
Data columns (total 50 columns):
year                      24624 non-null int64
player                    24624 non-null object
position                  24624 non-null object
age                       24624 non-null int64
team                      24624 non-null object
games                     24624 non-null int64
games_started             24624 non-null int64
minutes_played            24624 non-null int64
player_efficiency         24624 non-null float64
true_shooting_%           24624 non-null float64
three_pt_tries            24624 non-null float64
free_throws               24624 non-null float64
off_rebound_%             24624 non-null float64
def_rebound_%             24624 non-null float64
total_rebound_%           24624 non-null float64
assist_%                  24624 non-null float64
steal_%                   24624 non-null float64
block_%                   24624 non-null float64
turnover_%            

**Some research that may be useful in dealing with excessive NaNs in seasons_stats_df:**

- they started taking stats for offensive and defensive [rebounds](https://en.wikipedia.org/wiki/Rebound_(basketball)#:~:text=Both%20offensive%20and%20defensive%20rebounds,missed%20shots%20will%20likely%20land.) in 1973

- same year for [steals](https://en.wikipedia.org/wiki/Steal_(basketball)#:~:text=Steals%20were%20first%20recorded%20in,his%20on%20April%203%2C%201999.) and [blocks](https://en.wikipedia.org/wiki/List_of_National_Basketball_Association_annual_blocks_leaders#:~:text=The%20block%20title%20was%20first,on%20blocks%20were%20first%20compiled.)

- [turnover records](https://en.wikipedia.org/wiki/Turnover_(basketball)#:~:text=Turnovers%20were%20first%20officially%20recorded,during%20the%201967%E2%80%9368%20season.) in the NBA date back to 1977, but go all the way back in the ABA (American Basketball Association) to 1967 

- The four teams of the ABA (Denver Nuggets, Indiana Pacers, New York Nets, and San Antonio Spurs) [merged](https://www.washingtonpost.com/graphics/sports/nba-aba-merger/#:~:text=In%201976%2C%20four%20ABA%20teams,for%20the%20first%20three%20seasons.) with the NBA in 1976

In [80]:
# Attempted to count total NaN at each row in seasons_stats_df 
# By running the following code:

# for i in range(len(seasons_stats_df.index)) : 
#     print(" Total NaN in row", i + 1, ":", 
#           seasons_stats_df.iloc[i].isnull().sum()) 

# Realized after running this, it was going through and listing all NaNs
# found in each of the 24,691 rows individually.  And that, well, I don't have another 
# millions years to live.